packages/transaction-controller/src/TransactionController.ts

```diff
@@ -646,22 +624,111 @@ export class TransactionController extends BaseControllerV1<
     this.messagingSystem.subscribe(
       'NetworkController:stateChange',
       (_, patches) => {
-        if (this.#isMultichainEnabled) {
-          const networkClients = this.#getNetworkClientRegistry();
-          patches.forEach(({ op, path }) => {
-            if (op === 'remove' && path[0] === 'networkConfigurations') {
-              const networkClientId = path[1] as NetworkClientId;
-              delete networkClients[networkClientId];
-            }
-          });
+        // only needed to get `should not call getNetworkClientRegistry on networkController:stateChange when feature flag is disabled` to pass,
+        // otherwise can rely on the guard inside the helper
+        if (isMultichainEnabled) {
+        const networkClients = this.#getNetworkClientRegistry();
+        patches.forEach(({ op, path }) => {
+          if (op === 'remove' && path[0] === 'networkConfigurations') {
+            const networkClientId = path[1] as NetworkClientId;
+            delete networkClients[networkClientId];
+          }
+        });
 
-          this.#refreshTrackingMap(networkClients);
+        this.#multichainHelper.onNetworkClientsChange(networkClients);
         }
       },
     );
-    if (this.#isMultichainEnabled) {
-      this.#initTrackingMap();
-    }
+
+    this.#multichainHelper = new MultichainHelper({
+      isMultichainEnabled,
+      ethQuery: this.ethQuery,
+      nonceTracker: this.nonceTracker,
+      incomingTransactionOptions: incomingTransactions,
+      findNetworkClientIdByChainId: (chainId: Hex) => {
+        return this.messagingSystem.call(
+          `NetworkController:findNetworkClientIdByChainId`,
+          chainId,
+        );
+      },
+      getNetworkClientById: ((networkClientId: NetworkClientId) => {
+        return this.messagingSystem.call(
+          `NetworkController:getNetworkClientById`,
+          networkClientId,
+        )}) as NetworkController['getNetworkClientById'],
+      getNetworkClientRegistry,
+      removeIncomingTransactionHelperListeners:
+        this.#removeIncomingTransactionHelperListeners.bind(this),
+      removePendingTransactionTrackerListeners:
+        this.#removePendingTransactionTrackerListeners.bind(this),
+      createTrackersForNetworkClient:
+        this.#createTrackersForNetworkClient.bind(this),
+      acquireNonceLockForChainIdKey: this.#acquireNonceLockForChainIdKey.bind(this),
+    });
+  }
+
+  #createTrackersForNetworkClient(
+    networkClient: NetworkClient,
+    etherscanRemoteTransactionSource: EtherscanRemoteTransactionSource,
+  ) {
```

Comment:
any reason we cant move this to the helper too?

Reply:
We would have to pass a bunch of additional methods (`onConfirmedTransaction`, `setTransactionStatusDropped`, `failTransaction`, `updateTransaction` etc) into the helper as well then since they are bound as callbacks to the `pendingTransactionTracker` events in `addPendingTransactionTrackerListeners` and same for a couple of methods for `incomingTransactionHelper` events

Reply:
This feels like the core of what the `MultichainHelper` should be doing, creating the helper instances themselves.

If the issue is the number of dependencies that the helpers require, then rather than delegating the entire helper construction to the `TransactionController` and coupling it to the tracking, could we just have distinct constructor callbacks such as `createPendingTransactionTracker` and `createIncomingTransactionHelper`?

These wouldn't accept the full argument list, but only the arguments dictated by the `MultichainHelper` rather than anything the `TransactionController` already knows such as `getCurrentAccount` and the incoming transaction options.

Reply:
Done here https://github.com/MetaMask/core/pull/3896/commits/de7b5ac37b0842446ee8689acaec5b77dfee482e